### DISCUSSION TITLE
v7 custom service type check error? API changed?

### DISCUSSION THREAD

**Discussion by olivier-martin-sf** [webdriverio/webdriverio#6971] [Q&A]
Hi, I am updating one of our company project from v6 to v7, everything went pretty smoothly except for one last quirk. 

We use a custom in-house worker service in the config file. I moved the config file for `js`  to `ts` and when I use TS for type checking over the `wdio.conf.ts` I do have the following error:

> Type 'typeof IncCustomWdioService' is not assignable to type 'string'

Here's the service property value in the config file:

```ts
  services: [
    "chromedriver",
    [IncCustomWdioService, {}],
  ],
```

It seems that `services` as a `ServiceEntry[]` type and from what I read from the types in the wdio repo the values here seems legit as long as `IncCustomWdioService` is of type `ServiceClass`. 

Here's the Type declaration that correspond to `IncCustomWdioService`:

```ts
export declare class IncCustomWdioService {
    automationProtocol: string;
    driverConfig: DriverConfig;
    constructor(config: IncCustomWdioServiceConfig, capabilities: WdioCapabilities, wdioConfig: Config);
    before(capabilities: WdioCapabilities, specs: string[]): void;
}
```

So I was wondering if the APIs changed for authoring custom service or if it's something else. I didn't see anything related to that in the release note but if I compare the custom service doc from v6 to v7 it seems that some APIs changed (before seems to take a browser as the third parameter for instance). As I am not sure if it's a real issue or me misunderstanding the doc I prefer to ask here first.

Thanks in advance

**Comment by christian-bromann** [comment]
Hey @olivier-martin-sf,
thanks for filing the question. I wish I could transfer this into an issue because it seems we don't have enough documentation here. A service written in TS now needs to have the following structure:

```ts
import type { Services } from '@wdio/types'

export default class DevToolsService implements Services.ServiceInstance {
  // ...
}
```

This allows us to ensure that service parameters are compliant to the WebdriverIO package used. Can you try this with your custom service?

**Reply by olivier-martin-sf** [reply]
Hey @christian-bromann, thank you for your quick answer, I don't mind opening an issue if that would be valuable for you. I will give your suggestion a try and see if TS gets upset. I will keep you posted on my findings.
